File: src/automata/NFA.java
package automata;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NFA {

    private final String name;
    private final List<String> alphabet;
    private final Map<State, Map<String, List<State>>> graph;
    private final State initialState;
    private final List<State> finalStates;

    public NFA(String name, List<String> alphabet, Map<State, Map<String, List<State>>> graph, State initialState, List<State> finalStates) {
        this.name = name;
        this.alphabet = alphabet;
        this.graph = graph;
        this.finalStates = finalStates;
        this.initialState = initialState;
    }

    private boolean recursiveFunction(String w, State currentState) {

        if (w.isEmpty())
            return this.finalStates.contains(currentState);

        String target_edge = Character.toString(w.charAt(0));
        String restOf_w = w.substring(1);
        Map<String, List<State>> currentState_edges = this.graph.get(currentState);

        if (currentState_edges.containsKey(target_edge)) {

            List<State> possible_states = currentState_edges.get(target_edge);

            for (State item:possible_states) {
                if (recursiveFunction(restOf_w, item))
                    return true;
            }
        }
        else if (currentState_edges.containsKey("lambda")) {

            List<State> possible_states = currentState_edges.get("lambda");

            for (State item:possible_states) {
                if (recursiveFunction(w, item))
                    return true;
            }
        }

        return false;
    }

    public boolean doesAcceptString(String w) { return recursiveFunction(w, this.initialState); }

    public List<State> nextState(State st, String w) {
        List<State> states = new ArrayList<>();

        if (!w.isEmpty()) {
            if (this.graph.containsKey(st))
                if (this.graph.get(st).containsKey(w))
                    for (State item : this.graph.get(st).get(w)) {
                        states.add(item);
                        states.addAll(nextState(item, ""));
                    }
        }

        if (this.graph.containsKey(st))
            if (this.graph.get(st).containsKey("lambda"))
                for (State item:this.graph.get(st).get("lambda")) {
                    if (w.isEmpty())
                        states.add(item);
                    states.addAll(nextState(item, w));
                }

        return states;
    }

    private List<State> unionStates(List<State> states, String w) {

        List<State> union = new ArrayList<>();
        for (State item:states) {

            union.addAll(nextState(item, w));
        }

        return union;
    }

    private State CreateNewState(List<State> states) {

        if (states.isEmpty())
            return new State("empty_state");

        String state_name = "";
        for (State item:states) {
            state_name += item.getName() + "_";
        }

        return new State(state_name.substring(0,state_name.length()-1));
    }

    public DFA convertToDFA(){

        Map<List<State>, Map<String, List<State>>> graph = new HashMap<>();

        List<State> q0_state = new ArrayList<>();
        q0_state.add(this.initialState);

        Map<String, List<State>> q0_edges = new HashMap<>();
        for (String w:this.alphabet) {
            q0_edges.put(w, nextState(q0_state.get(0), w));
        }

        graph.put(q0_state, q0_edges);

        boolean graphIsComplete = false;

        while (!graphIsComplete) {
            graphIsComplete = true;
            Map<List<State>, Map<String, List<State>>> temp_graph = new HashMap<>();
            for (List<State> item:graph.keySet()) {
                Map<String, List<State>> edges = graph.get(item);
                for (List<State> jtem:edges.values())
                    if (!graph.containsKey(jtem)) {
                        graphIsComplete = false;

                        Map<String, List<State>> current_edges = new HashMap<>();
                        for (String w:this.alphabet) {

                            current_edges.put(w, unionStates(jtem, w));
                        }

                        temp_graph.put(jtem, current_edges);
                    }
            }

            for (List<State> item:temp_graph.keySet()) {
                graph.put(item, temp_graph.get(item));
            }
        }

        Map<State, Map<String, State>> dfa_graph = new HashMap<>();
        List<State> finals = new ArrayList<>();
        Map<List<State>, State> mapper = new HashMap<>();

        for (List<State> item:graph.keySet()) {

            State st;
            if (!mapper.containsKey(item)) {

                st = CreateNewState(item);
                for (State cst:item) {
                    if (this.finalStates.contains(cst)) {
                        finals.add(st);
                        break;
                    }
                }
                mapper.put(item, st);
            } else {

                st = mapper.get(item);
                for (State cst:item) {
                    if (this.finalStates.contains(cst)) {
                        finals.add(st);
                        break;
                    }
                }
            }

            Map<String, State> edges = new HashMap<>();

            for (String w:graph.get(item).keySet()) {
                List<State> states = graph.get(item).get(w);
                if (!mapper.containsKey(states)) {
                    State new_st = CreateNewState(states);
                    mapper.put(states, new_st);
                    edges.put(w, new_st);
                } else {
                    edges.put(w, mapper.get(states));
                }
            }
            dfa_graph.put(st, edges);
        }

        State init_state = new State("");
        for(State item:dfa_graph.keySet()) {
            if (item.getName().equals(this.initialState.getName()))
                init_state = item;
        }

        return new DFA(this.name + "_(DFA)", this.alphabet, dfa_graph, init_state, finals);
    }

    @Override
    public String toString() {

        return String.format("NFA name: %s\n graph:\n %s\n initial state: %s\n final states:\n %s\n",
                this.name, this.graph, this.initialState, this.finalStates);
    }
}

File: python_src/automata/nfa.py

from automata.state import State
from automata.dfa import DFA

class NFA:

    def __init__(self, name:str, comment:str, alphabet:list,
            graph:dict, initialState:State, finalStates:list):

        self._name = name
        self._comment = comment
        self._alphabet = frozenset(alphabet)
        self._graph = graph
        self._initialState = initialState
        self._finalStates = frozenset(finalStates)

    def recursiveFunction(self, w:str, currentState:State) -> bool:
        
        if not w:
            return currentState in self._finalStates

        target_edge = w[0]
        restOf_w = w[1:]
        currentState_edges = self._graph[currentState]

        if target_edge in currentState_edges.keys():

            possible_states = currentState_edges[target_edge]

            for item in possible_states:
                if self.recursiveFunction(restOf_w, item):
                    return True

        elif "lambda" in currentState_edges.keys():

            possible_states = currentState_edges["lambda"]

            for item in possible_states:
                if self.recursiveFunction(w, item):
                    return True

        return False

    def doesAcceptString(self, w:str) -> bool:
        return self.recursiveFunction(w, self._initialState)

    def nextState(self, st:State, w:str) -> frozenset:

        states = list()

        if w:
            if st in self._graph.keys():
                if w in self._graph[st].keys():
                    for item in self._graph[st][w]:
                        states.append(item)
                        for jtem in self.nextState(item, w[1:]):
                            states.append(jtem)

        if st in self._graph.keys():
            if "lambda" in self._graph[st].keys():
                for item in self._graph[st]["lambda"]:
                    if not w:
                        states.append(item)
                    for jtem in self.nextState(item, w):
                        states.append(jtem)

        return frozenset(states)

    def unionStates(self, states:frozenset, w:str) -> list:

        union = list()
        for item in states:
            for jtem in self.nextState(item, w):
                union.append(jtem)

        return frozenset(union)

    def createNewState(self, states:frozenset):
        
        if not states:
            return State("empty_state")

        state_name = "_".join([ x.getName() for x in states ])
        
        return State(state_name)

    def convertToDFA(self):

        graph = dict()

        q0_state = frozenset([self._initialState])
        q0_state = q0_state.union(self.nextState(self._initialState, "lambda"))

        q0_edges = dict()
        for i in self._alphabet:
            q0_edges[i] = self.unionStates(q0_state, i)

        graph[q0_state] = q0_edges
        q0_state_dfa = None

        graph_is_complete = False

        while not graph_is_complete:

            graph_is_complete = True
            temp_graph = dict()
            for item in graph.keys():
                edges = graph[item]

                for jtem in edges.values():
                    if jtem not in graph:
                        graph_is_complete = False

                        current_edges = dict()

                        for i in self._alphabet:
                            current_edges[i] = self.unionStates(jtem, i)

                        temp_graph[jtem] = current_edges



            for item in temp_graph.keys():
                graph[item] = temp_graph[item]

        dfa_graph = dict()
        finals = list()

        mapper = dict()
    
        for item in graph.keys():
            if item not in mapper.keys():
                st = self.createNewState(item)

                if item == q0_state:
                    q0_state_dfa = st

                for jtem in item:
                    if jtem in self._finalStates:
                        finals.append(st)
                        break
                
                mapper[item] = st

            else:
                
                st = mapper[item]
                for jtem in item:
                    if jtem in self._finalStates:
                        finals.append(st)
                        break

            edges = dict()

            for i in graph[item].keys():

                states = graph[item][i]
                if states not in mapper:
                    new_st = self.createNewState(states)
                    mapper[states] = new_st
                    edges[i] = new_st
                else:
                    edges[i] = mapper[states]

            dfa_graph[st] = edges

        return DFA(self._name + "_(DFA)", self._alphabet, dfa_graph, q0_state_dfa, frozenset(finals))
    
    def __str__(self):

        return "DFA name: %s\n comment: %s\n graph:\n %s\n initial state: %s\n final states:\n %s\n" % (self._name, self._comment, self._graph, self._initialState, list(self._finalStates))

    def __repr__(self):

        return str(self)

